### DISCUSSION TITLE
Clarification of AuthExchange docs.

### DISCUSSION THREAD

**Discussion by Undistraction** [urql-graphql/urql#1967] [Q&A] [answered]
Please could you clarify my understanding of how the options supplied to [AuthExchange](https://formidable.com/open-source/urql/docs/api/auth-exchange/) are used. 

I'm trying to understand how `didAuthError` and `willAuthError` are used in the sequence of events. As far as I can see, `willAuthError` provides a chance to proactively check any current credentials to see if a user needs to (re)authorise. I believe if it returns `true`, `getAuth` will be called before the request is sent, allowing reauthorisation. It also appears that `didAuthError` is called in the event of an error being returned from the request. In this case I believe that if it returns true, `getAuth` will be called before the request is retried. 

Is the above correct? If so, am I right in thinking the exchange will only allow either a proactive(`willAuthError`) _or_  a reactive (`didAuthError`) attempt at authorisation for each request (to avoid loops).

**Comment by JoviDeCroock** [comment, marked as answer]
The above is completely correct, willAuthError allows you to say "hey this token has expired" (or similar issues) and didAuthError allows you to observe an errored request and assess whether or not the auth was to blame for it. Both when returned a truthy value will call `getAuth`, willAuthError will defer the request and didAuthError will make it retry.

Both are allowed as we can have a client-side check and a reaction to a server-side event.

**Reply by Undistraction** [reply]
@JoviDeCroock Thank you very much.
